File: trunk/test/backend_test/src/com/jasonzqshen/familyAccountingBackendTest/entry/CreateWithTemplateTester.java
package com.jasonzqshen.familyAccountingBackendTest.entry;

import static org.junit.Assert.assertEquals;

import java.text.SimpleDateFormat;
import java.util.Date;

import com.jasonzqshen.familyAccountingBackendTest.utils.DocumentCreater;
import com.jasonzqshen.familyAccountingBackendTest.utils.MasterDataCreater;
import com.jasonzqshen.familyAccountingBackendTest.utils.TestData;
import com.jasonzqshen.familyAccountingBackendTest.utils.TestUtilities;
import com.jasonzqshen.familyAccountingBackendTest.utils.TesterBase;
import com.jasonzqshen.familyAccountingBackendTest.utils.TransactionDataChecker;
import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.document_entries.CustomerEntry;
import com.jasonzqshen.familyaccounting.core.document_entries.EntryTemplatesManagement;
import com.jasonzqshen.familyaccounting.core.document_entries.GLAccountEntry;
import com.jasonzqshen.familyaccounting.core.document_entries.VendorEntry;
import com.jasonzqshen.familyaccounting.core.transaction.DocumentIdentity;
import com.jasonzqshen.familyaccounting.core.transaction.HeadEntity;
import com.jasonzqshen.familyaccounting.core.transaction.TransactionDataManagement;

public class CreateWithTemplateTester extends TesterBase {
	EntryTemplatesManagement tempMgmt;

	@Override
	protected void doTest(CoreDriver coreDriver) throws Exception {
		TestUtilities.establishFolder2012_07(
				TestUtilities.TEST_ROOT_CREATE_WITH_TEMPLATE, coreDriver);
		tempMgmt = new EntryTemplatesManagement(coreDriver);
		//coreDriver.setRootPath(TestUtilities.TEST_ROOT_CREATE_WITH_TEMPLATE);
		assertEquals(true, coreDriver.isInitialized());

		MasterDataCreater.createMasterData(coreDriver);

		// create templates

		SimpleDateFormat format = new SimpleDateFormat("yyyy.MM.dd");
		// create document with template
		tempMgmt.initialize();
		TemplatesCreateTester.createTemplate(coreDriver, tempMgmt);

		Date date = format.parse("2012.07.02");

		VendorEntry vendor = (VendorEntry) tempMgmt.getEntryTemplate(1)
				.generateEntry();
		vendor.setValue(VendorEntry.POSTING_DATE, date);
		vendor.save(true);

		CustomerEntry customer = (CustomerEntry) tempMgmt.getEntryTemplate(3)
				.generateEntry();
		customer.setValue(CustomerEntry.POSTING_DATE, date);
		customer.setValue(CustomerEntry.AMOUNT, TestData.AMOUNT_CUSTOMER);
		customer.save(true);

		GLAccountEntry glEntry = (GLAccountEntry) tempMgmt.getEntryTemplate(2)
				.generateEntry();
		glEntry.setValue(GLAccountEntry.POSTING_DATE, date);
		glEntry.setValue(GLAccountEntry.AMOUNT, TestData.AMOUNT_GL);
		glEntry.save(true);

		TransactionDataManagement transManagement = coreDriver
				.getTransDataManagement();

		// month 08
		date = format.parse("2012.08.02");
		HeadEntity headEntity = DocumentCreater.createVendorDoc(coreDriver,
				date);
		DocumentIdentity docId = headEntity.getDocIdentity();
		transManagement.reverseDocument(docId);
	}

	@Override
	protected void check(CoreDriver coreDriver) throws Exception {
		TransactionDataChecker.checkTransactionData(coreDriver);

		// reload
		coreDriver.restart();
		TransactionDataChecker.checkTransactionData(coreDriver);

	}

}

File: proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingUnitTest/CreationTestCases/CreateWithTemplateTester.cs
using Microsoft.VisualStudio.TestPlatform.UnitTestFramework;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore;
using WinStore_FamilyAccountingCore.DocumentEntries;
using WinStore_FamilyAccountingCore.Transaction;
using WinStore_FamilyAccountingUnitTest.Utils;

namespace WinStore_FamilyAccountingUnitTest.CreationTestCases
{
    [TestClass]
    public class CreateWithTemplateTester
    {
        CoreDriver _coreDriver;
        [TestInitialize]
        public async Task Initialize()
        {
            _coreDriver = new CoreDriver();
            await StoragePreparation.EstablishFolder_2012_07(
                TestFolderPath.CREATING_TRANSDATA_WITH_TEMPLATE
                , _coreDriver);
        }

        [TestMethod]
        public async Task TestCreateTransWithTemplate()
        {
            Assert.IsTrue(_coreDriver.IsInitialize);
            MasterDataCreater.CreateMasterData(_coreDriver);

            await TemplateCreater.CreateTemplate(_coreDriver, _coreDriver.TmMgmt);

            // create transaction data
            DateTime date = DateTime.Parse("2012.07.02");

            VendorEntry vendor = (VendorEntry)_coreDriver.TmMgmt.GetEntryTemplate(1)
                    .GenerateEntry();
            vendor.SetValue(EntryTemplate.POSTING_DATE, date);
            await vendor.SaveAsync(true);

            CustomerEntry customer = (CustomerEntry)_coreDriver.TmMgmt.GetEntryTemplate(3)
                    .GenerateEntry();
            customer.SetValue(EntryTemplate.POSTING_DATE, date);
            customer.SetValue(EntryTemplate.AMOUNT, TestData.AMOUNT_CUSTOMER);
            await customer.SaveAsync(true);

            GLAccountEntry glEntry = (GLAccountEntry)_coreDriver.TmMgmt.GetEntryTemplate(2)
                    .GenerateEntry();
            glEntry.SetValue(EntryTemplate.POSTING_DATE, date);
            glEntry.SetValue(EntryTemplate.AMOUNT, TestData.AMOUNT_GL);
            await glEntry.SaveAsync(true);

            TransactionDataManagement transManagement = _coreDriver.TransMgmt;

            // month 08
            date = DateTime.Parse("2012.08.02");
            HeadEntity headEntity = await DocumentCreater.CreateVendorDoc(_coreDriver,
                    date);
            DocumentIdentity docId = headEntity.DocIdentity;
            transManagement.ReverseDocument(docId);

            // check
            TransactionDataChecker.CheckTransactionData(_coreDriver);

            // reload
            await _coreDriver.RestartAsync();
            TransactionDataChecker.CheckTransactionData(_coreDriver);
        }
    }
}
